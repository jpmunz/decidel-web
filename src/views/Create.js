import React from "react";
import { Alert, Form } from "react-bootstrap";
import ListInput from "./ListInput";
import { getAsyncRequest, createDecidelRequest } from "src/state";
import LoadingButton from "./LoadingButton";
import { Redirect } from "react-router-dom";
import { connect } from "react-redux";
import autoBind from "react-autobind";
import "./Create.scss";

export class Create extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
    autoBind(this);
  }

  handleNameChange(event) {
    this.setState({ name: event.target.value });
  }

  handleTitleChange(event) {
    this.setState({ title: event.target.value });
  }

  handleOptionsChange(options) {
    this.setState({ options: options });
  }

  handleDecidersChange(deciders) {
    this.setState({ deciders: deciders });
  }

  handleCreate(event) {
    event.preventDefault();

    const deciders = [{ name: this.state.name }];
    this.state.deciders.forEach((name, i) => {
      deciders.push({ name: name, isNext: i === 0 });
    });

    this.props.createDecidelRequest({
      requestId: this.props.requestId,
      decidel: {
        deciders: deciders,
        title: this.state.title,
        options: this.state.options.map(option => ({ title: option })),
        history: []
      }
    });
  }

  isValid() {
    return (
      this.state.name &&
      this.state.name.length &&
      this.state.title &&
      this.state.title.length &&
      this.state.deciders &&
      this.state.deciders.length &&
      this.state.deciders.every(item => item.length) &&
      this.state.options &&
      this.state.options.length > 1 &&
      this.state.options.every(item => item.length)
    );
  }

  render() {
    if (this.props.request.success) {
      return <Redirect to={"/view/" + this.props.request.response.id} />;
    }

    let alert;
    if (this.props.request.error) {
      alert = (
        <Alert variant="warning">
          There was a problem creating your Decidel. Please try again later.
        </Alert>
      );
    }

    return (
      <Form className="Create">
        <Form.Group>
          <Form.Label htmlFor="name-input">What's your name?</Form.Label>
          <Form.Control
            id="name-input"
            disabled={this.props.request.inProgress}
            onChange={this.handleNameChange}
            type="text"
            placeholder="The Decider"
          />
        </Form.Group>

        <Form.Group>
          <Form.Label htmlFor="title-input">
            What are you trying to decide?
          </Form.Label>
          <Form.Control
            id="title-input"
            disabled={this.props.request.inProgress}
            onChange={this.handleTitleChange}
            type="text"
            placeholder="What movie to watch"
          />
        </Form.Group>

        <Form.Group>
          <Form.Label>What options do you have?</Form.Label>
          <ListInput
            disabled={this.props.request.inProgress}
            numInitialOptions={2}
            onChange={this.handleOptionsChange}
            addMessage="Add another option"
          />
        </Form.Group>

        <Form.Group>
          <Form.Label>Who else is deciding?</Form.Label>
          <ListInput
            disabled={this.props.request.inProgress}
            onChange={this.handleDecidersChange}
            addMessage="Add another decider"
          />
        </Form.Group>

        <div className="footer">
          {alert}
          <LoadingButton
            className="create-button"
            type="submit"
            loading={this.props.request.inProgress}
            disabled={!this.isValid()}
            onClick={this.handleCreate}
            variant="primary"
            text="Create"
          />
        </div>
      </Form>
    );
  }
}

export default connect(
  (state, props) => ({
    request: getAsyncRequest(state, props.requestId)
  }),
  { createDecidelRequest }
)(Create);
