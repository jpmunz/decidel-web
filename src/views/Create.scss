@import "mixins";

.Create {
  width: 60%;
  margin: 50px auto;
  padding: 10px;
  @include rounded();
  @include box-shadow(10px, 10px, 20px, rgba(0, 0, 0, 0.2));

  .form-control {
    border-top: none;
    border-left: none;
    border-right: none;
    border-radius: 0;
    padding: 0;
    box-shadow: none;
    width: calc(100% - 40px);
  }

  .form-group {
    margin: 0 auto;
    width: 85%;
  }

  .form-label {
    margin-top: 30px;
    margin-bottom: 20px;
  }

  .footer {
    text-align: right;
  }

  .create-button {
    @include box-shadow(5px, 5px, 10px, rgba(0, 0, 0, 0.2));
  }
}
