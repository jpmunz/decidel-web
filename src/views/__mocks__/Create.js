import React from "react";

export const MockCreate = jest.fn();

MockCreate.mockReturnValue(<div>Mocked Create</div>);

const mock = jest.fn().mockImplementation(MockCreate);

export default mock;
