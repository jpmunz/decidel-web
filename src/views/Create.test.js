import React from "react";
import { Create } from "./Create";
import { render, fireEvent } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

const createMock = jest.fn();

const renderComponent = function({ requestId = "22", request = {} }) {
  return render(
    <BrowserRouter>
      <Create
        requestId={requestId}
        request={request}
        createDecidelRequest={createMock}
      />
    </BrowserRouter>
  );
};

const renderTest = options =>
  expect(renderComponent(options).asFragment()).toMatchSnapshot();

test("Create form submission", () => {
  const component = renderComponent({ requestId: "22" });

  fireEvent.change(component.getByLabelText("What's your name?"), {
    target: { value: "Jon" }
  });
  fireEvent.change(component.getByLabelText("What are you trying to decide?"), {
    target: { value: "What to Eat" }
  });

  const listInputs = component.getAllByTestId("list-input-option");
  fireEvent.change(listInputs[0], { target: { value: "Fish" } });
  fireEvent.change(listInputs[1], { target: { value: "Veggies" } });
  fireEvent.change(listInputs[2], { target: { value: "Bob" } });

  fireEvent.click(component.getByText("Create"));

  expect(createMock.mock.calls).toEqual([
    [
      {
        decidel: {
          title: "What to Eat",
          deciders: [{ name: "Jon" }, { name: "Bob", isNext: true }],
          history: [],
          options: [{ title: "Fish" }, { title: "Veggies" }]
        },
        requestId: "22"
      }
    ]
  ]);
});

test("Blank create form", () => renderTest({}));
test("Create form submitting", () =>
  renderTest({ request: { inProgress: true } }));
test("Create form submission success", () =>
  renderTest({ request: { success: true, response: { id: "a1" } } }));
test("Create form submission error", () =>
  renderTest({ request: { error: "failed" } }));
