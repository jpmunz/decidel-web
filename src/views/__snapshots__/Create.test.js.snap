// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blank create form 1`] = `
<DocumentFragment>
  <form
    class="Create"
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="name-input"
      >
        What's your name?
      </label>
      <input
        class="form-control"
        id="name-input"
        placeholder="The Decider"
        type="text"
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="title-input"
      >
        What are you trying to decide?
      </label>
      <input
        class="form-control"
        id="title-input"
        placeholder="What movie to watch"
        type="text"
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
      >
        What options do you have?
      </label>
      <ul
        class="ListInput"
      >
        <li>
          <div
            class="input-group"
          >
            <input
              class="form-control"
              data-testid="list-input-option"
              type="text"
              value=""
            />
            <div
              class="remove-button-container input-group-append"
            >
              <button
                class="ImageButton btn btn-primary"
                tabindex="-1"
                type="button"
              >
                <img
                  alt="remove"
                  src="remove.svg"
                />
              </button>
            </div>
          </div>
        </li>
        <li>
          <div
            class="input-group"
          >
            <input
              class="form-control"
              data-testid="list-input-option"
              type="text"
              value=""
            />
            <div
              class="remove-button-container input-group-append"
            >
              <button
                class="ImageButton btn btn-primary"
                tabindex="-1"
                type="button"
              >
                <img
                  alt="remove"
                  src="remove.svg"
                />
              </button>
            </div>
          </div>
        </li>
        <li>
          <input
            alt="add-another"
            class="add-another form-control"
            placeholder="Add another option"
            type="text"
            value=""
          />
        </li>
      </ul>
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
      >
        Who else is deciding?
      </label>
      <ul
        class="ListInput"
      >
        <li>
          <div
            class="input-group"
          >
            <input
              class="form-control"
              data-testid="list-input-option"
              type="text"
              value=""
            />
            <div
              class="remove-button-container input-group-append"
            >
              <button
                class="ImageButton btn btn-primary"
                tabindex="-1"
                type="button"
              >
                <img
                  alt="remove"
                  src="remove.svg"
                />
              </button>
            </div>
          </div>
        </li>
        <li>
          <input
            alt="add-another"
            class="add-another form-control"
            placeholder="Add another decider"
            type="text"
            value=""
          />
        </li>
      </ul>
    </div>
    <div
      class="footer"
    >
      <button
        class="create-button btn btn-primary"
        disabled=""
        type="submit"
      >
        Create
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`Create form submission error 1`] = `
<DocumentFragment>
  <form
    class="Create"
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="name-input"
      >
        What's your name?
      </label>
      <input
        class="form-control"
        id="name-input"
        placeholder="The Decider"
        type="text"
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="title-input"
      >
        What are you trying to decide?
      </label>
      <input
        class="form-control"
        id="title-input"
        placeholder="What movie to watch"
        type="text"
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
      >
        What options do you have?
      </label>
      <ul
        class="ListInput"
      >
        <li>
          <div
            class="input-group"
          >
            <input
              class="form-control"
              data-testid="list-input-option"
              type="text"
              value=""
            />
            <div
              class="remove-button-container input-group-append"
            >
              <button
                class="ImageButton btn btn-primary"
                tabindex="-1"
                type="button"
              >
                <img
                  alt="remove"
                  src="remove.svg"
                />
              </button>
            </div>
          </div>
        </li>
        <li>
          <div
            class="input-group"
          >
            <input
              class="form-control"
              data-testid="list-input-option"
              type="text"
              value=""
            />
            <div
              class="remove-button-container input-group-append"
            >
              <button
                class="ImageButton btn btn-primary"
                tabindex="-1"
                type="button"
              >
                <img
                  alt="remove"
                  src="remove.svg"
                />
              </button>
            </div>
          </div>
        </li>
        <li>
          <input
            alt="add-another"
            class="add-another form-control"
            placeholder="Add another option"
            type="text"
            value=""
          />
        </li>
      </ul>
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
      >
        Who else is deciding?
      </label>
      <ul
        class="ListInput"
      >
        <li>
          <div
            class="input-group"
          >
            <input
              class="form-control"
              data-testid="list-input-option"
              type="text"
              value=""
            />
            <div
              class="remove-button-container input-group-append"
            >
              <button
                class="ImageButton btn btn-primary"
                tabindex="-1"
                type="button"
              >
                <img
                  alt="remove"
                  src="remove.svg"
                />
              </button>
            </div>
          </div>
        </li>
        <li>
          <input
            alt="add-another"
            class="add-another form-control"
            placeholder="Add another decider"
            type="text"
            value=""
          />
        </li>
      </ul>
    </div>
    <div
      class="footer"
    >
      <div
        class="fade alert alert-warning show"
        role="alert"
      >
        There was a problem creating your Decidel. Please try again later.
      </div>
      <button
        class="create-button btn btn-primary"
        disabled=""
        type="submit"
      >
        Create
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`Create form submission success 1`] = `
<DocumentFragment>
  <div>
    Mocked Redirect
  </div>
</DocumentFragment>
`;

exports[`Create form submitting 1`] = `
<DocumentFragment>
  <form
    class="Create"
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="name-input"
      >
        What's your name?
      </label>
      <input
        class="form-control"
        disabled=""
        id="name-input"
        placeholder="The Decider"
        type="text"
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="title-input"
      >
        What are you trying to decide?
      </label>
      <input
        class="form-control"
        disabled=""
        id="title-input"
        placeholder="What movie to watch"
        type="text"
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
      >
        What options do you have?
      </label>
      <ul
        class="ListInput"
      >
        <li>
          <div
            class="input-group"
          >
            <input
              class="form-control"
              data-testid="list-input-option"
              disabled=""
              type="text"
              value=""
            />
            <div
              class="remove-button-container input-group-append"
            >
              <button
                class="ImageButton btn btn-primary"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <img
                  alt="remove"
                  src="remove.svg"
                />
              </button>
            </div>
          </div>
        </li>
        <li>
          <div
            class="input-group"
          >
            <input
              class="form-control"
              data-testid="list-input-option"
              disabled=""
              type="text"
              value=""
            />
            <div
              class="remove-button-container input-group-append"
            >
              <button
                class="ImageButton btn btn-primary"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <img
                  alt="remove"
                  src="remove.svg"
                />
              </button>
            </div>
          </div>
        </li>
        <li>
          <input
            alt="add-another"
            class="add-another form-control"
            disabled=""
            placeholder="Add another option"
            type="text"
            value=""
          />
        </li>
      </ul>
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
      >
        Who else is deciding?
      </label>
      <ul
        class="ListInput"
      >
        <li>
          <div
            class="input-group"
          >
            <input
              class="form-control"
              data-testid="list-input-option"
              disabled=""
              type="text"
              value=""
            />
            <div
              class="remove-button-container input-group-append"
            >
              <button
                class="ImageButton btn btn-primary"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <img
                  alt="remove"
                  src="remove.svg"
                />
              </button>
            </div>
          </div>
        </li>
        <li>
          <input
            alt="add-another"
            class="add-another form-control"
            disabled=""
            placeholder="Add another decider"
            type="text"
            value=""
          />
        </li>
      </ul>
    </div>
    <div
      class="footer"
    >
      <button
        class="create-button btn btn-primary"
        disabled=""
        type="submit"
      >
        <span
          aria-hidden="true"
          class="spinner-border spinner-border-sm"
          role="status"
        />
        <span
          class="sr-only"
        >
          Loading...
        </span>
      </button>
    </div>
  </form>
</DocumentFragment>
`;
